com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs

```diff
@@ -358,21 +358,54 @@ public IReadOnlyList<ulong> ConnectedClientsIds
+        /// Connection Approval Result
         /// </summary>
-        /// <param name="createPlayerObject">If true, a player object will be created. Otherwise the client will have no object.</param>
-        /// <param name="playerPrefabHash">The prefabHash to use for the client. If createPlayerObject is false, this is ignored. If playerPrefabHash is null, the default player prefab is used.</param>
-        /// <param name="approved">Whether or not the client was approved</param>
-        /// <param name="position">The position to spawn the client at. If null, the prefab position is used.</param>
-        /// <param name="rotation">The rotation to spawn the client with. If null, the prefab position is used.</param>
-        public delegate void ConnectionApprovedDelegate(bool createPlayerObject, uint? playerPrefabHash, bool approved, Vector3? position, Quaternion? rotation);
+        /// <param name="Approved">Whether or not the client was approved</param>
+        /// <param name="CreatePlayerObject">If true, a player object will be created. Otherwise the client will have no object.</param>
+        /// <param name="PlayerPrefabHash">The prefabHash to use for the client. If createPlayerObject is false, this is ignored. If playerPrefabHash is null, the default player prefab is used.</param>
+        /// <param name="Position">The position to spawn the client at. If null, the prefab position is used.</param>
+        /// <param name="Rotation">The rotation to spawn the client with. If null, the prefab position is used.</param>
+        public struct ConnectionApprovalResult
+        {
+            public bool Approved;
+            public bool CreatePlayerObject;
+            public uint? PlayerPrefabHash;
+            public Vector3? Position;
+            public Quaternion? Rotation;
+        }
+
+        /// <summary>
+        /// Connection Approval Request
+        /// </summary>
+        /// <param name="Payload">The connection data payload</param>
+        /// <param name="ClientNetworkId">The Client Network Id of the client we are about to approve</param>
+        public struct ConnectionApprovalRequest
+        {
+            public byte[] Payload;
+            public ulong ClientNetworkId;
+        }
 
         /// <summary>
-        /// The callback to invoke during connection approval
+        /// The callback to invoke during connection approval. Allows client code to decide to allow a client in or not
         /// </summary>
-        public event Action<byte[], ulong, ConnectionApprovedDelegate> ConnectionApprovalCallback = null;
+        public Func<ConnectionApprovalRequest, ConnectionApprovalResult> ConnectionApprovalHandler
+        {
+            get => m_ConnectionApprovalHandler;
+            set
+            {
+                if (value != null && value.GetInvocationList().Length > 1)
+                {
+                    Debug.LogError(
```

Comment:
throw an `InvalidOperationException` instead?

Reply:
s/instead/additionally ?

Reply:
an unhandled exception is eventually an error/exception log in Unity console anyway

Reply:
so still, instead

Reply:
What I mean to say is that `throw (new InvalidOperationException());` will be less informational than the log. See updated code. But ok, no big deal, I can remove the log too.

Reply:
what about `throw new InvalidOperationException($"Only one connection approval handler is supported in {nameof(m_ConnectionApprovalCallback)}. Rejecting further adds.");` ? :)

Reply:
One, good, we should be good to go now. Please let me know if I missed any bits or parts.